ansible/roles/netplan/defaults/main.yml

```diff
@@ -0,0 +1,115 @@
+
+# .. Copyright (C) 2015-2017 Maciej Delmanowski <drybjed@gmail.com>
+# .. Copyright (C) 2015-2017 Robin Schneider <ypid@riseup.net>
+# .. Copyright (C) 2015-2017 DebOps <https://debops.org/>
+# .. SPDX-License-Identifier: GPL-3.0-only
+
+# .. _ifupdown__ref_defaults:
+
+# debops.ifupdown default variables [[[
+# =====================================
+
+# .. contents:: Sections
+#    :local:
+#
+# .. include:: ../../../../includes/global.rst
+
+
+# APT packages [[[
+# ----------------
+
+# .. envvar:: netplan__base_packages [[[
+#
+# List of base APT packages to install.
+netplan__base_packages:
+  - [ 'nplan', 'netplan.io' ]
+
+                                                                   # ]]]
+# .. envvar:: netplan__packages [[[
+#
+# List of additional APT packages to install with the role.
+netplan__packages: []
+                                                                   # ]]]
+
+# .. envvar:: netplan__purge_packages [[[
+#
+# List of APT packages which will be purged when this role is
+# enabled, to stop them from interfering with ``netplan``.
+netplan__purge_packages: []
+                                                                   # ]]]
+                                                                   # ]]]
+
+# General role configuration [[[
+# ------------------------------
+
+# .. envvar:: netplan__reconfigure_auto [[[
+#
+# If ``True``, the role will reconfigure interfaces automatically after changes
+# are detected by Ansible. If ``False``, role will install the reconfiguration
+# script on the remote host so that the user can run it to apply the changes by
+# hand.
+netplan__reconfigure_auto: True
+
+                                                                   # ]]]
+# .. envvar:: netplan__config_file [[[
+#
+# Full path to the netplan config file to deploy
+netplan__config_file: '/etc/netplan/ansible-config.yaml'
+
+                                                                   # ]]]
+# .. envvar:: netplan__remove_existing_configs [[[
+#
+# If ``True``, the role will remove every existing config files in /etc/netplan/
+netplan__remove_existing_config: False
+
+                                                                   # ]]]
+# .. envvar:: netplan__renderer [[[
+#
+# Choices are:
+#
+# * ``networkd``
+# * ``NetworkManager``
+#
+netplan__renderer: networkd
+
+                                                                   # ]]]
+# Netplan configuration [[[
+# -----------------------------------
+
+# .. envvar:: neplan__configuration [[[
+#
+# Dictionary with the netplan config to deploy e.g.
+#
+# netplan__configuration:
+#   network:
+#     version: 2
+#     ethernets:
+#       eth0:
+#         addresses:
+#           - 192.168.2.22/24
+#         nameservers:
+#           addresses: [192.168.2.1]
+#         routes:
+#           - to: default
+#             via: 192.168.2.1
+#       eth1:
+#         addresses:
+#           - 192.168.34.3/24
+
+netplan__configuration: {}
```

Comment:
I would urge you to get familiar with DebOps [Universal Configuration](https://docs.debops.org/en/stable-3.0/user-guide/universal-configuration.html) and use it in this role. Since `netplan` uses YAML for its configuration, a good starting point would be to check out other DebOps roles that deploy YAML configuration. For example the `lxd` role has "preseed" configuration done in YAML using Universal Configuration.

A default configuration defined in the role can be useful - it's usually the default Ethernet interface with DHCP enabled.

Reply:
Personally, in sensitive roles like this one which reconfigure the network and can leave the machine inaccesible if a wrong config is applied I prefer to avoid providing a default configuration to make sure that the applied configuration has been defined by the user.

That's also why I apply this conditional before applying the netplan network config

```
  when:
    - netplan__configuration.keys() | length > 0
```

With the current implementation any netplan configuration can be applied using the inventory without modifying the role itself. 

If using universal configuration is mandatory I will look at it

Reply:
I wouldn't call it "mandatory" but encouraged. :-) The problem it can solve here is that the configuration template is strict with bridge, bond, vlan entries. If `netplan` extends this configuration, any new features will have to be explicitly added to the template. With Universal Configuration implementation, you can add any new features in the inventory and the role doesn't need to be updated.

For a first iteration of the role, going with a YAML dictionary is fine though, no complaints. If `netplan` can be installed and doesn't complain about default configuration being present, going with no default seems OK with me as well. But in this case, I expect extensive documentation with multiple examples. :-)